### PR TITLE
[BEAM-8539] Make job state transitions in python-based runners consistent with java-based runners

### PR DESCRIPTION
**Pull request by chadrik**
Update the python SDK to properly reflect that STOPPED is not a terminal state.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by chadrik** [comment]
Run Java PreCommit

**Comment by chadrik** [comment]
R: @lukecwik 
R: @mxm 
R: @robertwb 

This is a followup to https://github.com/apache/beam/pull/9969 that corrects python-based runners to behave the same as Java-based runners wrt STOPPED: it is the initial state and not a terminal state.  

Tests pass, but I'm not sure if this could cause problems with Dataflow.

File: sdks/python/apache_beam/runners/portability/local_job_service.py [hunk 1]
```diff
@@ -195,7 +195,7 @@ def __init__(self,
     self._state = None
     self._state_queues = []
     self._log_queues = []
-    self.state = beam_job_api_pb2.JobState.STARTING
+    self.state = beam_job_api_pb2.JobState.STOPPED
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/local_job_service.py]: Why this change?

File: model/job-management/src/main/proto/beam_job_api.proto [hunk 2]
```diff
@@ -213,16 +213,37 @@ message JobMessagesResponse {
...
+
+    // The job has completed draining its data. (terminal)
     DRAINED = 8;
+
+    // The job is starting up.
     STARTING = 9;
+
+    // The job is cancelling.
     CANCELLING = 10;
     UPDATING = 11;
```
Comment (robertwb) [inline model/job-management/src/main/proto/beam_job_api.proto]: No comment on UPDATING?

Comment (chadrik) [reply model/job-management/src/main/proto/beam_job_api.proto]: Ah, sorry. that was added by @lukecwik after I created this.  Will do.

File: sdks/python/apache_beam/runners/portability/local_job_service.py (hunk 1 above)
Comment (chadrik) [reply sdks/python/apache_beam/runners/portability/local_job_service.py]: The initial state in java runners is STOPPED.  Then it transitions to STARTING and RUNNING.  More info at #9969.

**Comment by chadrik** [comment]
> Should STOPPED be PAUSED?

Based on the comments describing STOPPED, I'd say yes.  But @lukecwik has informed me in the other PR that there are no runners that actually support being paused (RUNNING -> STOPPED -> RUNNING transition) , so STOPPED is not currently used this way.

File: model/job-management/src/main/proto/beam_job_api.proto (hunk 2 above)
Comment (lukecwik) [reply model/job-management/src/main/proto/beam_job_api.proto]: `// The job is in the progress of being updated.`

File: model/job-management/src/main/proto/beam_job_api.proto [hunk 3]
```diff
@@ -213,16 +213,37 @@ message JobMessagesResponse {
 // without needing to pass through STARTING.
 message JobState {
   enum Enum {
+    // The job state reported by a runner cannot be interpreted by the SDK.
     UNSPECIFIED = 0;
+
+    // The job has been paused, or has not yet started.
     STOPPED = 1;
+
+    // The job is currently running. (terminal)
```
Comment (lukecwik) [inline model/job-management/src/main/proto/beam_job_api.proto]: not terminal?

File: model/job-management/src/main/proto/beam_job_api.proto [hunk 4]
```diff
@@ -213,16 +213,37 @@ message JobMessagesResponse {
...
+    // The job has successfully completed. (terminal)
     DONE = 3;
+
+    // The job has failed. (terminal)
     FAILED = 4;
+
+    // The job has been explicitly cancelled. (terminal)
     CANCELLED = 5;
+
+    // The job has been updated.
```
Comment (lukecwik) [inline model/job-management/src/main/proto/beam_job_api.proto]: terminal?

File: model/job-management/src/main/proto/beam_job_api.proto (hunk 3 above)
Comment (chadrik) [reply model/job-management/src/main/proto/beam_job_api.proto]: whoops! fixed.

File: model/job-management/src/main/proto/beam_job_api.proto (hunk 4 above)
Comment (chadrik) [reply model/job-management/src/main/proto/beam_job_api.proto]: fixed

File: model/job-management/src/main/proto/beam_job_api.proto (hunk 2 above)
Comment (chadrik) [reply model/job-management/src/main/proto/beam_job_api.proto]: fixed

**Comment by chadrik** [comment]
Run Python PreCommit

**Comment by chadrik** [comment]
Run Java PreCommit

**Comment by chadrik** [comment]
Review notes have been addressed.

File: model/job-management/src/main/proto/beam_job_api.proto [hunk 5]
```diff
@@ -213,17 +213,40 @@ message JobMessagesResponse {
 // without needing to pass through STARTING.
 message JobState {
   enum Enum {
+    // The job state reported by a runner cannot be interpreted by the SDK.
     UNSPECIFIED = 0;
+
+    // The job has been paused, or has not yet started.
```
Comment (robertwb) [inline model/job-management/src/main/proto/beam_job_api.proto]: Didn't we decide that there is no pause/restart? 

Stopped really sounds like the job was running at one time, until someone decided to stop it. Perhaps this would be worth renaming to UNSTARTED?

Comment (chadrik) [reply model/job-management/src/main/proto/beam_job_api.proto]: I removed "paused" from the description.  I agree that the name STOPPED is misleading, but renaming it would be a much bigger task that could leak into non-portable runners.  Hopefully the improvement surrounding the documentation of these enums will suffice for now.
